# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

导入数据分析所需要的库，并通过Pandas的`read_csv`函数，将原始数据文件"e_commerce.csv"里的数据内容，解析为DataFrame，并赋值给变量`original_data`。

In [1]:
import pandas as pd

In [55]:
import os
print(os.getcwd())

C:\Users\杨加春\data_analysis


读取数据与写入文件的注意点：
- 由上面的os.getcwd()可知当前jupyter文件是存放于C盘users目录下的data_analysis中的，学过python文件路径你可知如果你在data_analysis文件中存在所要读取的文件你可以直接用该文件名读取
- 如果要读取的文件在其它地方那么你就需要用绝对路径来读取
- 写入时填写路径时你如果用文件名当作路径那么会在data_analysis文件夹下建立文件，如果用绝对路径指定要写入的文件才会把内容写入其它任意位置的文件中
- 若要写入的文件不存在会自建一个，若存在则会清空该文件原内容后再写入

In [4]:
original_data = pd.read_csv(r"F:\BaiduNetdiskDownload\数据分析课程操作的文件\04数据评估与清洗篇\e_commerce.csv")

In [5]:
original_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 评估数据

在这一部分，我将对在上一部分建立的`original_data`这个DataFrame所包含的数据进行评估。

评估主要从两个方面进行：结构和内容，即整齐度和干净度。数据的结构性问题指不符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”这三个标准，数据的内容性问题包括存在丢失数据、重复数据、无效数据等

### 评估数据整齐度

In [6]:
original_data.sample(10)

InvoiceNo StockCode                         Description  Quantity  \
434261    574035     21666         RIDGED GLASS T-LIGHT HOLDER         2   
328363    565767     20728                 LUNCH BAG CARS BLUE        15   
497318    578362     22153     ANGEL DECORATION STARS ON DRESS        48   
369619    569060     23401       RUSTIC MIRROR WITH LACE HEART         2   
435112    574074     21990        MODERN FLORAL STATIONERY SET         3   
131662    547580    85123A  WHITE HANGING HEART T-LIGHT HOLDER        12   
293762    562616     23308  SET OF 60 VINTAGE LEAF CAKE CASES         24   
164133    550636     22807        SET OF 6 T-LIGHTS TOADSTOOLS         1   
165500    550804     21714        CITRONELLA CANDLE GARDEN POT        12   
90477     544040     22841        ROUND CAKE TIN VINTAGE GREEN         1   

             InvoiceDate  UnitPrice  CustomerID         Country  
434261   11/2/2011 12:45       0.65     14606.0  United Kingdom  
328363    9/6/2011 14:28       4.96         NaN  United Kingdom  
497318  11/24/2011 10:39       0.42     12913.0  United Kingdom  
369619   9/30/2011 10:52       6.25     13954.0  United Kingdom  
435112   11/2/2011 15:33       2.46         NaN  United Kingdom  
131662   3/24/2011 10:48       2.95     13488.0  United Kingdom  
293762    8/8/2011 12:28       0.55     14748.0  United Kingdom  
164133   4/19/2011 15:37       5.79         NaN  United Kingdom  
165500   4/20/2011 15:32       1.25     16676.0  United Kingdom  
90477    2/15/2011 11:40       7.95     14701.0  United Kingdom

从抽样的10行数据数据来看，数据符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”，具体来看每行是关于某商品的一次交易，每列是交易相关的各个变量，因此不存在结构性问题。

### 评估数据干净度

In [7]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


从输出结果来看，数据共有541909条观察值，而`Description`、`CustomerID`变量存在缺失值。

此外，`InvoiceDate`的数据类型应为日期，`CustomerID`的数据类型应为字符串，应当进行数据格式转换。

#### 评估缺失数据

在了解`Description`存在缺失值后，根据条件提取出缺失观察值。

In [8]:
original_data[original_data["Description"].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

有1454条交易数据缺失`Description`变量值。

从输出结果来看，这些缺失`Description`的交易数据，`UnitPrice`都为0。为了验证猜想，我们增加筛选条件，看是否存在`Description`变量缺失且`UnitPrice`不为0的数据。

In [9]:
original_data[(original_data["Description"].isnull()) & (original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

筛选出来结果数量为0条，说明缺失`Description`值的数据，同时也不具备有效的`UnitPrice`值。

`Description`表示产品名称，`UnitPrice`表示产品单价，都是进行后续商品交易分析的重要变量。如果它们同时缺失/无效，我们认为数据无法提供有效含义，因此这些后续可以被删除。

`CustomerID`变量同样存在缺失值，因此也根据条件提取出缺失观察值。

In [17]:
print(original_data["CustomerID"].isnull())
type(original_data["CustomerID"].isnull())

0         False
1         False
2         False
3         False
4         False
          ...  
541904    False
541905    False
541906    False
541907    False
541908    False
Name: CustomerID, Length: 541909, dtype: bool


pandas.core.series.Series

In [10]:
original_data[original_data["CustomerID"].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

In [ ]:
`CustomerID`表示客户编号，不是分析畅销商品的必要变量。并且从输出结果来看，有些`CustomerID`缺失的交易数据仍然有效，因此保留此变量为空的观察值。

#### 评估重复数据

根据数据变量的含义来看，虽然`InvoiceNo`、`StockCode`和`CustomerID`都是唯一标识符，但一次交易可能包含多件商品，因此`InvoiceNo`可以存在重复，不同交易可以包含同件商品，因此`StockCode`可以存在重复，顾客可以进行多次交易或下单多个商品，因此`CustomerID`也可以存在重复。

那么针对此数据集，我们无需评估重复数据。

#### 评估不一致数据

不一致数据可能存在于`Country`变量中，我们要查看是否存在多个不同值指代同一国家的情况。

In [23]:
print(type(original_data["Country"].value_counts()))
original_data["Country"].value_counts()

<class 'pandas.core.series.Series'>


Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


从`Country`变量值来看，`"USA"`、`"United States"`均在表示美国，`"United Kingdom"`、`"UK"`、`"U.K."`均在表示英国，因此应该在清洗步骤对这些值进行统一，只保留一个指代值。

#### 评估无效或错误数据

可以通过DataFrame的`describe`方法，对数值统计信息进行快速了解。

In [13]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

从输出结果来看，`Quantity`和`UnitPrice`存在负数，会对后续数值分析产生影响。

因此，我们先筛选出`Quantity`数值为负数的观察值，进一步评估其含义。

In [20]:
print(type(original_data["Quantity"] < 0))
original_data["Quantity"] < 0

<class 'pandas.core.series.Series'>


0         False
1         False
2         False
3         False
4         False
          ...  
541904    False
541905    False
541906    False
541907    False
541908    False
Name: Quantity, Length: 541909, dtype: bool

In [21]:
original_data[original_data["Quantity"] < 0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

从筛选结果来看，`Quantity`变量为负数的观察值，`InvoiceNo`似乎均以C开头，表示订单被取消。

但此猜测不一定准确，为了验证猜想，我们增加筛选条件，看是否存在`Quantity`变量为负数且`InvoiceNo`不以C开头的观察值。

**你会发现对df的列调用相关方法如：.isnull()/.value_counts()或一些判定 >0/!=0pandas会返回一个相对应的series**

In [22]:
type(original_data["InvoiceNo"].str[0] != "C")

pandas.core.series.Series

In [24]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C")]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

In [ ]:
以上猜想被证实错误，因为还存在`Quantity`变量为负数且`InvoiceNo`不以C开头的观察值。

但根据以上输出结果，这些筛选出的观察值的`UnitPrice`观察值均为0，因此增加`UnitPrice`的条件进行验证。

In [25]:
original_data[(original_data["Quantity"] < 0)&(original_data["InvoiceNo"].str[0] != "C")&(original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

根据输出结果，猜想得到验证，当`Quantity`变量为负数的时候，观察值满足以下条件之一：
- `InvoiceNo`以C开头，表示订单被取消
- `UnitPrice`为0，说明单价为0英镑

这些交易数据均不是有效成交数据，不贡献销售，不在后续分析范围内，因此我们将在数据清理步骤，将`Quantity`为负数的观察值删除。

接下来筛选出`UnitPrice`为负数的观察值。

In [26]:
original_data[original_data["UnitPrice"] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

从输出结果来看，`UnitPrice`为负数的观察值都是坏账调账(description)，不属于实际商品交易数据，因此也在数据清理步骤中也将其删除。

## 清理数据

根据前面评估部分得到的结论，我们需要进行的数据清理包括：
- 把`InvoiceDate`变量的数据类型转换为为日期时间
- 把`CustomerID`变量的数据类型转换为字符串
- 把`Description`变量缺失的观察值删除
- 把`Country`变量值`"USA"`替换为`"United States"`
- 把`Country`变量值`"UK"`、`"U.K."`替换为`"United Kingdom"`
- 把`Quantity`变量值为负数的观察值删除
- 把`UnitPrice`变量值为负数的观察值删除

为了区分开经过清理的数据和原始的数据，我们创建新的变量`cleaned_data`，让它为`original_data`复制出的副本。我们之后的清理步骤都将被运用在`cleaned_data`上。

>在python以及基于python的库/架构中'变量1=变量2'赋值要注意：b =[1,2,3],a = b这只是给同一个数据对象起了一个新别名即：a,b均指向列表[1,2,3]之后你对a/b操作那么结果都会反映在列表[1,2,3]上而 .copy() 才是真正地创建了一个全新的、独立的副本a = b.copy(),此时Python 会在内存中开辟一块全新的空间，将b的内容完整地复制过去，然后让a指向这个新的空间之后对a/b的操作将互不干扰

In [30]:
cleaned_data = original_data.copy()
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [ ]:
把`InvoiceDate`变量的数据类型转换为为日期时间类型：

In [31]:
cleaned_data["InvoiceDate"] = pd.to_datetime(cleaned_data["InvoiceDate"])
cleaned_data["InvoiceDate"]

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

把`CustomerID`变量的数据类型转换为字符串：

In [34]:
type(cleaned_data["CustomerID"].astype(str))

pandas.core.series.Series

In [35]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].astype(str)
cleaned_data["CustomerID"]

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

**注意str/int/float都是object的子类，所以转换为str后显示为object类型也对**

把`CustomerID`变量值结尾的`".0"`删除：

In [36]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].str.slice(0, -2)
cleaned_data["CustomerID"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

把`Description`变量缺失的观察值删除，并查看删除后该列空缺值个数和：

In [41]:
cleaned_data.dropna(subset=["Description"], inplace=True)

**注意：python中True/False等价于数值1/0并进行数值运算而isnull返回一个全为True/False的series对其用.sum求和最终数值就是True-nan的个数**

In [42]:
cleaned_data["Description"].isnull().sum()

np.int64(0)

把`Country`变量值`"USA"`替换为`"United States"`，并检查替换后`"USA"`变量值个数：

In [43]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"USA": "United States"})

In [44]:
len(cleaned_data[cleaned_data["Country"] == "USA"])

0

把`Country`变量值`"UK"`、`"U.K."`替换为`"United Kingdom"`，并检查替换后`"UK"`和`"U.K."`变量值个数：

In [46]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"UK": "United Kingdom", "U.K.": "United Kingdom"})


In [47]:
print(len(cleaned_data[cleaned_data["Country"] == "UK"]))
print(len(cleaned_data[cleaned_data["Country"] == "U.K."]))

0
0


把`Quantity`变量值为负数的观察值删除，并检查替换后`Quantity`变量值为负数的个数：

In [48]:
cleaned_data = cleaned_data[cleaned_data["Quantity"] >= 0]

*对已存在的变量再赋值会改变该变量值<br>用已存在的变量赋值新变量会使两个变量均指向同一个数据对象 <br>对不存在的列(变量)赋值会创造新列,对已存在的列赋值会改变原值*

In [49]:
len(cleaned_data[cleaned_data["Quantity"] < 0])

0

把`UnitPrice`变量值为负数的观察值删除，并检查替换后`UnitPrice`变量值为负数的个数：

In [50]:
cleaned_data = cleaned_data[cleaned_data["UnitPrice"] >= 0]

In [51]:
len(cleaned_data[cleaned_data["UnitPrice"] < 0])

0

## 保存清理后的数据

完成数据清理后，把干净整齐的数据保存到新的文件里，文件名为`e_commerce_cleaned.csv`。

In [52]:
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice CustomerID         Country  
0 2010-12-01 08:26:00       2.55      17850  United Kingdom  
1 2010-12-01 08:26:00       3.39      17850  United Kingdom  
2 2010-12-01 08:26:00       2.75      17850  United Kingdom  
3 2010-12-01 08:26:00       3.39      17850  United Kingdom  
4 2010-12-01 08:26:00       3.39      17850  United Kingdom

In [53]:
cleaned_data.to_csv("e_commerce_cleaned_SavedInDataAnalysis.csv", index=False)

In [54]:
pd.read_csv("e_commerce_cleaned_SavedInDataAnalysis.csv").head()

C:\Users\杨加春\AppData\Local\Temp\ipykernel_17400\764538907.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv("e_commerce_cleaned_SavedInDataAnalysis.csv").head()


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice CustomerID         Country  
0  2010-12-01 08:26:00       2.55      17850  United Kingdom  
1  2010-12-01 08:26:00       3.39      17850  United Kingdom  
2  2010-12-01 08:26:00       2.75      17850  United Kingdom  
3  2010-12-01 08:26:00       3.39      17850  United Kingdom  
4  2010-12-01 08:26:00       3.39      17850  United Kingdom

>注意在pandas的写入与读取时：写入时若不指定index=False，pandas默认会把自带的下标索引0123存到写入的文件中<br>但当你读取存好的文件时pandas会把那个下标索引列当作一个属性列显示在df上又因为该列没设属性名会自动为其配备untitled0属性名并再次为其分配下标索引0123<br>所以每次把清理好的df存入文件之前若df采取默认下标0123当作索引则一定要设置index=False如果df已经用某列当作标签索引没有用默认的0123下标那么就无需设置index=False

**df/series.str会让Pandas返回一个用于对 Series（DataFrame 的某一列）中的每个字符串进行操作的方法集,df.str.slice() 则是调用对字符串的切片操作（左闭右开）slice(start, stop, step-默认为1)**
